File: Conversoes de Bases/Decimal e Octal/DecimalOctal.lua
function DecimalParaOctal(decimal)
  local octal = ""
  while decimal > 0 do
    local resto = decimal % 8
    octal = resto .. octal
    decimal = math.floor(decimal / 8)
  end
  return tonumber(octal)
end

function OctalParaDecimal(octal)
  local decimal = 0
  local potencia = 0
  while octal > 0 do
    local digito = octal % 10
    decimal = decimal + digito * (8 ^ potencia)
    octal = math.floor(octal / 10)
    potencia = potencia + 1
  end
  return math.floor(decimal)
end

function IsValidOctal(octal)
  while octal > 0 do
    if octal % 10 > 7 then
      return false
    end
    octal = math.floor(octal / 10)
  end
  return true
end

local LimiteMaximoDecimal = 2147483647
local LimiteMaximoOctal = 17777777777

repeat
  print("=========================")
  print(" CONVERSOR DECIMAL-OCTAL ")
  print("=========================")
  print()
  print("1 - Converter decimal para octal")
  print("2 - Converter octal para decimal")
  print("0 - Sair")
  print()
  io.write("Digite a opcao desejada: ")
  local opcao = tonumber(io.read())
  if opcao == 1 then
    print()
    io.write("Digite um numero decimal (ate ", LimiteMaximoDecimal, "): ")
    local decimal = tonumber(io.read())
    if decimal > LimiteMaximoDecimal then
      print("Valor decimal fora do limite permitido.")
    else
      print("Valor em octal: " .. DecimalParaOctal(decimal))
    end
  elseif opcao == 2 then
    print()
    io.write("Digite um numero octal (ate ", LimiteMaximoOctal, "): ")
    local octal = tonumber(io.read())
    if not IsValidOctal(octal) then
      print("Valor octal invalido.")
    elseif octal > LimiteMaximoOctal then
      print("Valor octal fora do limite permitido.")
    else
      print("Valor em decimal: " .. OctalParaDecimal(octal))
    end
  end
  print()
until opcao == 0

File: Conversoes de Bases/Decimal e Octal/DecimalOctal.pas
program DecimalOctal;

uses
  SysUtils, Math;

const
  LimiteMaximoDecimal: Cardinal = 2147483647;
  LimiteMaximoOctal: Cardinal = 17777777777;

function DecimalParaOctal(decimal: Cardinal): Cardinal;
var
  resto: Cardinal;
  potencia: Cardinal;
  octal: Cardinal;
begin
  octal := 0;
  potencia := 1;
  while decimal > 0 do
  begin
    resto := decimal mod 8;
    octal := octal + resto * potencia;
    decimal := decimal div 8;
    potencia := potencia * 10;
  end;
  DecimalParaOctal := octal;
end;

function OctalParaDecimal(octal: Cardinal): Cardinal;
var
  potencia: Cardinal;
  digito: Cardinal;
  decimal: Cardinal;
begin
  decimal := 0;
  potencia := 0;
  while octal > 0 do
  begin
    digito := octal mod 10;
    decimal := decimal + digito * Trunc(Power(8, potencia));
    octal := octal div 10;
    potencia := potencia + 1;
  end;
  OctalParaDecimal := decimal;
end;

function IsValidOctal(octal: Cardinal): Boolean;
begin
  while octal > 0 do
  begin
    if (octal mod 10) > 7 then
    begin
      IsValidOctal := False;
      Exit;
    end;
    octal := octal div 10;
  end;
  IsValidOctal := True;
end;

var
  opcao: Integer;
  decimal: Cardinal;
  octal: Cardinal;

begin
  repeat
    Writeln('=========================');
    Writeln(' CONVERSOR DECIMAL-OCTAL ');
    Writeln('=========================');
    Writeln;
    Writeln('1 - Converter decimal para octal');
    Writeln('2 - Converter octal para decimal');
    Writeln('0 - Sair');
    Writeln;
    Write('Digite a opcao desejada: ');
    Readln(opcao);
    case opcao of
      1:
      begin
        WriteLn;
        Write('Digite um numero decimal (ate ', LimiteMaximoDecimal, '): ');
        Readln(decimal);
        if decimal > LimiteMaximoDecimal then
          Writeln('Valor decimal fora do limite permitido.')
        else
          Writeln('Valor em octal: ', DecimalParaOctal(decimal));
      end;
      2:
      begin
        WriteLn;
        Write('Digite um numero octal (ate ', LimiteMaximoOctal, '): ');
        Readln(octal);
        if not IsValidOctal(octal) then
          Writeln('Valor octal invalido.')
        else if octal > LimiteMaximoOctal then
          Writeln('Valor octal fora do limite permitido.')
        else
          Writeln('Valor em decimal: ', OctalParaDecimal(octal));
      end;
    end;
    Writeln;
  until opcao = 0;
end.
